src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.Generated.cs

```diff
@@ -12,6 +14,24 @@ partial class LambdaCompiler
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Maintainability", "CA1505:AvoidUnmaintainableCode"), System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Maintainability", "CA1502:AvoidExcessiveComplexity")]
         private void EmitExpression(Expression node, CompilationFlags flags)
         {
+            // When compling deep trees, we run the risk of triggering a terminating StackOverflowException,
+            // so we probe for sufficient stack here. When we run out of stack, we simply continue the work
+            // on another thread. Note that EnsureSufficientExecutionStack is very fast, so the overhead of
+            // this overflow protection is minimal.
```

Comment:
Pushed an update with a `StackGuard` helper class that encapsulates the recursion depth tracking, probing after reaching a threshold, and the `RunOnEmptyStack` helper methods. It also has a `MaxExecutionStackCount` to limit the number of stacks we create; we can either put this threshold high or remove it altogether if we're confident that recursive trees cannot be constructed.

As a data point, the test case of a tree with depth 10000 requires a stack count of > 32 (but < 64) to get properly compiled by the `LightCompiler`. The `LambdaCompiler` seemed to survive with a lower threshold. Upon quick inspection, it seems we could save some stack storage if we change methods like:

```csharp
        private void CompileBinaryExpression(Expression expr)
        {
            var node = (BinaryExpression)expr;
```

to take in a strongly typed expression type, and perform the cast in the dispatch method higher up. That'd save us a local variable slot and prolong the time to reach the insufficient stack condition. I'll open a separate issue to track this possibility.